src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -1289,70 +1281,72 @@ private static unsafe bool TryInt32ToHexStr(int value, char hexBase, int digits,
             }
 
             charsWritten = bufferLength;
-            fixed (char* buffer = &MemoryMarshal.GetReference(destination))
-            {
-                char* p = Int32ToHexChars(buffer + bufferLength, (uint)value, hexBase, digits);
-                Debug.Assert(p == buffer);
-            }
+
+            int p = Int32ToHexChars(destination.Slice(0, bufferLength), (uint)value, hexBase, digits);
+            Debug.Assert(p == 0);
+
             return true;
         }
 
-        private static unsafe char* Int32ToHexChars(char* buffer, uint value, int hexBase, int digits)
+        private static int Int32ToHexChars(Span<char> buffer, uint value, int hexBase, int digits)
         {
-            while (--digits >= 0 || value != 0)
+            int bufferEndIndex = buffer.Length;
+            while ((--digits >= 0 || value != 0) && /* Only to elide bounds check */ (uint)(--bufferEndIndex) < (uint)buffer.Length)
```

Comment:
Did this actually elide the bounds checks and did it insert additional comparisons for the loop condition?

Reply:
It add extra instructions, but overall the code is smaller: [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHufuAHKAEsAbtgwwm5JA0EA7DAwCS8uqQAqEABIwELABbYouABQBlPtlkAeMAagA+BsACuAM1cwoaBs7kLRADbOMN5+DHo6AEJ4ITLyDAAmggDmghi4AJRcvDyc1DkFcQou7p4AorIJygk6DAC8Tm4eUAB0ADIwsskYegDc2YU8AO56ggESxsZwcEmp6Qz2DQAMDISEDIHBDACEyxkMAGQHDAD0AFQMAPKyAQCeDBgQDDABgjVOEM6VuAy2MGAAawYZxODGMvnkGSmcBKzQqVUqOn2VjBEIwGVhnnanW6eiy+UGuQGhN4wFu4kSKTS9TBZPEUM2EmOSwQADEMv0CSTSU1PABtTFQeHVHQAXRpxlshihs2pAGowbKFCjyCsAPy08kwDIAciWOoYIHCURiGQ5xO5G2wQQk9kWDBQnMtAF8LYViAB2RqlIWVEUIJ2FV1chhu/hCUQU4hSIpKFTqLQ6fSGXAAfUuAQSZgs1ilDm9zW8aKtNtC8QiCGiuFiYSVmTDHAbvDCguFiIQNMF2K6PUDJJGYwm0LrC2Wq3WjJ2e3xlryltJWspcwldO1xknzLZ5pDlsFAt5voRNQQ4oakrsMqpCgVxiVDBV6s19L1BqNFar2u3854k7tDUdTY8MGlqegW5R+u2fY5MGzpAA)

Reply:
It would be better to move the `(uint)(--bufferEndIndex) < (uint)buffer.Length` into an `if` that surrounds the loop.
This keeps the bounds elision while also keeping the check outside of the loop and so improves the overall perf with minimal impact (rather than incurring it per iteration).
Same with the other case above.

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHufuAHKAEsAbtgwwm5JA0EA7DAwCS8uqQAqEABIwELABbYouABQBlPtlkAeMAagA+BsACuAM1cwoaBs7kLRADbOMN5+DHo6AEJ4ITLyDAAmggDmghi4AJRcvDyc1DkFcQou7p4AorIJygk6DAC8Tm4eUAB0ADIwsskYegDc2YU8AO56ggESxsZwcEmp6Qz2DQAMDISEDIHBDACEyxkMAGQHDAD0AFQMAPKyAQCeDBgQDDABgjVOEM6VuAy2MGAAawYZxODGMvnkGSmcBKzQqVUqOn2VjBEIwGVhnnanW6eiy+UGuQGhN4wFu4kSKTS9TBZPEUM2EmOSwQADEMv0CSTSU1PABtTFQeHVHQAXRpxlshihs2pAGowbKFCjyCsAPy08kwDIAciWOoYIHCURiGQ5xO5G2wQQk9kWDBQnMtAF8LYViAB2RqlIWVEUIJ2FV1chhu/hCUQU4hSIpKFTqLQ6fSGXCkMwWaxShze5reNFWm2heIRBDRXCxMJKzJhjg13hhQXCxEIGmC7FdHqBkmCVxg8F+KHTRt+5vI1EDtsdDt4utEkOWkZjCbTKsLZardaMnZ7WcFPKWwZ0iRKiVHhnWrbMtnm+cH7iCgW830ImoIcUNSV2GVUhQK4wnlV1U1ek9QNI0SzLbUbzvAotztBpHV3HJgxdJDPRzcoR1fLtkJoZ0gA==

Reply:
--bufferEndIndex need to be in the while loop, to decrease its value every iteration

Reply:
Ah right and the modification in the loop causes the JIT to lose track of the bounds elision. This seems good enough for now then and I'll log a bug to track us removing it.

I'd just like us to get as close to "same perf" as possible (which looks to mostly be ensuring the bounds checks stay gone) so we don't regress.